discovery/openstack/instance.go

```diff
@@ -0,0 +1,211 @@
+	// OpenStack API reference
+	// https://developer.openstack.org/api-ref/compute/#list-floating-ips
+	pagerFIP := floatingips.List(client)
+	floatingIPList := make(map[string][]string)
+	err = pagerFIP.EachPage(func(page pagination.Page) (bool, error) {
+		result, err := floatingips.ExtractFloatingIPs(page)
+		if err != nil {
+			return false, fmt.Errorf("could not extract floatingips: %s", err)
+		}
+		for _, ip := range result {
+			// Skip not associated ips
+			if ip.InstanceID != "" {
+				floatingIPList[ip.InstanceID] = append(floatingIPList[ip.InstanceID], ip.IP)
+			}
+		}
+		return true, nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not describe floating IPs: %s", err)
+	}
+
+	// OpenStack API reference
+	// https://developer.openstack.org/api-ref/compute/#list-servers
+	opts := servers.ListOpts{}
+	pager := servers.List(client, opts)
+	tg := &config.TargetGroup{
+		Source: fmt.Sprintf("OS_" + i.region),
+	}
+	err = pager.EachPage(func(page pagination.Page) (bool, error) {
+		instanceList, err := servers.ExtractServers(page)
+		if err != nil {
+			return false, fmt.Errorf("could not extract instances: %s", err)
+		}
+
+		for _, s := range instanceList {
+			labels := model.LabelSet{
+				openstackLabelInstanceID: model.LabelValue(s.ID),
+			}
+			if len(s.Addresses) == 0 {
+				i.logger.Info("Got no IP address for instance %s", s.ID)
+				continue
+			}
+			for _, address := range s.Addresses {
+				md, ok := address.([]interface{})
+				if !ok {
+					i.logger.Warn("Invalid type for address, expected array")
+					continue
+				}
+				if len(md) == 0 {
+					i.logger.Debugf("Got no IP address for instance %s", s.ID)
+					continue
+				}
+				md1, ok := md[0].(map[string]interface{})
+				if !ok {
+					i.logger.Warn("Invalid type for address, expected dict")
+					continue
+				}
+				addr, ok := md1["addr"].(string)
+				if !ok {
+					i.logger.Warn("Invalid type for address, expected string")
+					continue
+				}
+				labels[openstackLabelPrivateIP] = model.LabelValue(addr)
+				addr = net.JoinHostPort(addr, fmt.Sprintf("%d", i.port))
+				labels[model.AddressLabel] = model.LabelValue(addr)
+				// Only use first private IP
+				break
+			}
+			if val, ok := floatingIPList[s.ID]; ok && len(val) > 0 {
+				labels[openstackLabelPublicIP] = model.LabelValue(val[0])
+			}
+			labels[openstackLabelInstanceStatus] = model.LabelValue(s.Status)
+			labels[openstackLabelInstanceName] = model.LabelValue(s.Name)
+			id, ok := s.Flavor["id"].(string)
+			if !ok {
+				i.logger.Warn("Invalid type for instance id, excepted string")
+				continue
+			}
+			labels[openstackLabelInstanceFlavor] = model.LabelValue(id)
+			for k, v := range s.Metadata {
+				name := strutil.SanitizeLabelName(k)
+				labels[openstackLabelTagPrefix+model.LabelName(name)] = model.LabelValue(v)
+			}
+			tg.Targets = append(tg.Targets, labels)
+		}
+		return true, nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not describe instances: %s", err)
```

Comment:
Same thing, `return nil, err` would simplify the generated error message.

Reply:
Updated.